homeassistant/components/rituals_perfume_genie/sensor.py

```diff
@@ -0,0 +1,168 @@
+        entities.append(DiffuserFillSensor(diffuser, coordinator))
+        entities.append(DiffuserWifiSensor(diffuser, coordinator))
+        if BATTERY in diffuser.data[HUB][SENSORS]:
+            entities.append(DiffuserBatterySensor(diffuser, coordinator))
+
+    async_add_entities(entities)
+
+
+class DiffuserPerfumeSensor(DiffuserEntity):
+    """Representation of a diffuser perfume sensor."""
+
+    def __init__(self, diffuser, coordinator):
+        """Initialize the perfume sensor."""
+        super().__init__(diffuser, coordinator, PERFUME_SUFFIX)
+
+    @property
+    def icon(self):
+        """Return the perfume sensor icon."""
+        if self.coordinator.data[HUB][SENSORS][PERFUME][ID] == PERFUME_NO_CARTRIDGE_ID:
+            return "mdi:tag-remove"
+        return "mdi:tag-text"
+
+    @property
+    def state(self):
+        """Return the state of the perfume sensor."""
+        return self.coordinator.data[HUB][SENSORS][PERFUME][TITLE]
+
+
+class DiffuserFillSensor(DiffuserEntity):
+    """Representation of a diffuser fill sensor."""
+
+    def __init__(self, diffuser, coordinator):
+        """Initialize the fill sensor."""
+        super().__init__(diffuser, coordinator, FILL_SUFFIX)
+
+    @property
+    def icon(self):
+        """Return the fill sensor icon."""
+        if self.coordinator.data[HUB][SENSORS][FILL][ID] == FILL_NO_CARTRIDGE_ID:
+            return "mdi:beaker-question"
+        return "mdi:beaker"
+
+    @property
+    def state(self):
+        """Return the state of the fill sensor."""
+        return self.coordinator.data[HUB][SENSORS][FILL][TITLE]
+
+
+class DiffuserBatterySensor(DiffuserEntity):
+    """Representation of a diffuser battery sensor."""
+
+    def __init__(self, diffuser, coordinator):
+        """Initialize the battery sensor."""
+        super().__init__(diffuser, coordinator, BATTERY_SUFFIX)
+
+    @property
+    def state(self):
+        """Return the state of the battery sensor."""
+        # Use ICON because TITLE may change in the future.
+        # ICON filename does not match the image.
+        return {
+            "battery-charge.png": 100,
+            "battery-full.png": 100,
+            "battery-75.png": 50,
+            "battery-50.png": 25,
+            "battery-low.png": 10,
+        }[self.coordinator.data[HUB][SENSORS][BATTERY][ICON]]
+
+    @property
+    def _charging(self):
+        """Return battery charging state."""
+        return bool(
+            self.coordinator.data[HUB][SENSORS][BATTERY][ID] == BATTERY_CHARGING_ID
+        )
+
+    @property
+    def device_class(self):
+        """Return the class of the battery sensor."""
+        return DEVICE_CLASS_BATTERY
+
+    @property
+    def extra_state_attributes(self):
+        """Return the battery state attributes."""
+        return {
+            ATTR_BATTERY_LEVEL: self.coordinator.data[HUB][SENSORS][BATTERY][TITLE],
+            ATTR_BATTERY_CHARGING: self._charging,
```

Comment:
Charging on/off should be a separate binary sensor. Please remove the extra state attributes.